examples/pytorch/seal_ogbl/main.py

```diff
@@ -0,0 +1,426 @@
+        self.root = root
+        self.graph = graph
+        self.split = split
+        self.split_edge = split_edge
+        self.num_hops = num_hops
+        self.percent = percent
+        self.ratio_per_hop = ratio_per_hop
+        self.directed = directed
+        self.dynamic = dynamic
+
+        if 'weights' in self.graph.edata:
+            self.edge_weights = self.graph.edata['weights']
+        else:
+            self.edge_weights = None
+        if 'feat' in self.graph.ndata:
+            self.node_features = self.graph.ndata['feat']
+        else:
+            self.node_features = None
+
+        if not self.dynamic:
+            self.examples = self.load_cached()
+            return
+
+        pos_edge, neg_edge = get_pos_neg_edges(split, self.split_edge, self.graph, self.percent)
+        self.links = torch.cat([pos_edge, neg_edge], 0).tolist() # [Np + Nn, 2]
+        self.labels = [1] * len(pos_edge) + [0] * len(neg_edge)
+
+    def __len__(self):
+        return len(self.links)
+
+    def __getitem__(self, idx):
+        if not self.dynamic:
+            return self.examples[idx]
+
+        src, dst = self.links[idx]
+        y = self.labels[idx]
+        subg = k_hop_subgraph(src, dst, self.num_hops, self.graph,
+            self.ratio_per_hop, self.directed)
+
+        # remove the link between src and dst
+        if subg.has_edges_between(0, 1):
+            subg.remove_edges(subg.edge_ids(0, 1))
+        if subg.has_edges_between(1, 0):
+            subg.remove_edges(subg.edge_ids(1, 0))
+
+        NIDs, EIDs = subg.ndata[dgl.NID], subg.edata[dgl.EID]
+
+        z = drnl_node_labeling(subg.adj(scipy_fmt='csr'), 0, 1)
+        edge_weights = self.edge_weights[EIDs] if self.edge_weights is not None else None
+        x = self.node_features[NIDs] if self.node_features is not None else None
+
+        subg_aug = subg.add_self_loop()
+        if edge_weights is not None:
+            edge_weights = torch.cat([
+                edge_weights, torch.ones(subg_aug.num_edges() - subg.num_edges())])
+        return subg_aug, z, x, edge_weights, y
+
+    @property
+    def cached_name(self):
+        return 'SEAL_{}_data_{}.pt'.format(self.split, self.percent)
+
+    def process(self):
+        examples = []
+        self.dynamic = True
+        for i in range(len(self)):
+            examples.append(self[i])
+        self.dynamic = False
+        return examples
+
+    def load_cached(self):
+        path = os.path.join(self.root, self.cached_name)
+        if os.path.exists(path):
+            return torch.load(path)
+
+        if not os.path.exists(self.root):
+            os.makedirs(self.root)
+
+        pos_edge, neg_edge = get_pos_neg_edges(
+            self.split, self.split_edge, self.graph, self.percent)
+        self.links = torch.cat([pos_edge, neg_edge], 0).tolist() # [Np + Nn, 2]
+        self.labels = [1] * len(pos_edge) + [0] * len(neg_edge)
+
+        examples = self.process()
+        torch.save(examples, path)
```

Comment:
It is recommended to save DGLGraphs with `save_graphs` and load them with `load_graphs`. See https://docs.dgl.ai/api/python/dgl.html#utilities